basic oewa support

hjanuschka opened the pull request:
This one try's to add basic oewa support.
Oewa is a derivation of the infoonline - as it seems.



you need to host a helper-html on a different subdomain (via https)

```
<html>
  <head>
    <script src=https://dispatcher.oewabox.at/oewa.js></script>
  </head>
  <body>
    <script>
      var match,
          pl     = /\+/g,
          search = /([^&=]+)=?([^&]*)/g,
          decode = function (s) { return decodeURIComponent(s.replace(pl, " ")); },
          query  = window.location.search.substring(1);
      var OEWA = {};
      while (match = search.exec(query)) { OEWA[decode(match[1])] = decode(match[2]); }
      oewa.ns = true;
      oewa.c({}, OEWA, 2);
    </script>
  </body>
</html>
```



call it with:
```
<!-- OEWA example -->
<amp-analytics type="oewa">
<script type="application/json">
{
  "vars": {
    "s": "offer",
    "cp": "RedCont/Nachrichten/NachrichtenUeberblick"
  },
  "requests": {
    "url": "https://subdomain.domain.example/amp-analytics-oewa.html"
  }
}
</script>
</amp-analytics>
<!-- End OEWA example -->
```


i am not a oewa-official - i am trying to get one of them on board too - and let them validate the implementation.

**my only problem is that the `${url}` from the vendors config gets encoded, is there anyway i can disable the encoding on the one variable?**


>log.js:246 Uncaught (in promise) Error: undefined source must start with "https://" or "//" or be relative and served from either https or from localhost. Invalid value: https://https%3A%2F%2Fsecure.krone.at%2Famp_oewa.html?site=ssl-krone&cp=%2F…ulterschluss!_TTIP_bei_uns_so_gut_wie_tot-Leaks_als_Sargnagel-Story-508282​​​(…)

hjanuschka commented:
I signed it!

dvoytenko commented:
/cc @cramforce @avimehta

avimehta commented:
Hi @hjanuschka Can you explain the reason why iframe is needed to send the data? 

Fir the encoding problem, you need to put the https:// in the requests section. Keep just the domain and path in var.

hjanuschka commented:
@avimehta thx for the help.

  * rebased current master
  * squased to single commit
  * added sample to `analytics.amp.html`
  * tests should pass


i am going to forward this PR to one of the oewa officials - and waiting for feedback - if we can merge this.

avimehta commented:
@hjanuschka: Can you let me know when you hear back from oewa?

hjanuschka commented:
there needs to be a board-decision - should happen mid-may. i am trying to keep these PR green and mergeable - and reach you once we have a decision.
thx

hjanuschka commented:
ok - updated:
  * rebased current master
  * made changes according the pr's from @oewaoffice

oewaoffice commented:
LGTM

hjanuschka commented:
@avimehta as we got the LGTM now can you merge and proceed it?

cramforce commented:
@hjanuschka You will need to get a LGTM from a project member. This PR cannot be merged until the implementation is updated in accordance with my comment on #3379

hjanuschka commented:
@cramforce  sorry to bother you  - i meant the LGTM from the oewa-official's - sorry

i think this is not an easy one - as most of the publishers in austria require to have oewa implementation (for myself we at krone.at can't launch amp articles w/o oewa) ,i hope it will be merged - and later on oewa itself should try to get rid of localstorage and use amp-uid (but i guess it takes a bit of time to replace localstorage)


so like `infoonline`, `adobeanalytics` it should be handled with the iframe exception!

ping @cramforce

cramforce commented:
Right, but at the very least we need to get their plan first. If their script relies on `localStorage` it will simply not work at all in iOS, so unless you are fine with Android only tracking, you will need to go the full way.

The examples you listed that currently use the exception went through the same process and have removal of the exception scheduled.

oewaoffice commented:
I'm afraid, according to the ÖWA measurement policy the iframe embed must be on a publisher's domain for a valid count. Changes of measurement policies require consent in the steering committee and takes usually several months.

cramforce commented:
I took a look at the request string generated on krone.at:
http://krone.oewabox.at/j0=,,,r=http%3A%2F%2Fwww.krone.at%2F;+,cp=RedCont%2FNachrichten%2FNachrichtenueberblick%2Fmoewa%2F+ssl=0+hn=mobil.krone.at;;;?lt=iowyki5d&x=412x732x24

Shouldn't it be straight forward to simply construct this string using `amp-analytics`?

How about this:

```
"oewa": {
 +    "pageview": "https://krone.oewabox.at/j0=,,,r=${canonicalUrl};+,cp=${title}%2Fmoewa%2F+ssl=1+hn=${canonicalHost};;;?lt=${pageViewId}&x=${screenWidth}x${screenHeight}x24"
 +  },
```

I think that should get you pretty far.

Cheers
Malte

cramforce commented:
@oewaoffice See solution above. Should work fine.

@hjanuschka Would you mind testing and updating the pull request?

hjanuschka commented:
@cramforce  my first try - was like this - but with this they have to rely on cookie based Unique-Client detection - wich has been superseeded by localstorage.

as from my debugs, it seems that ios has a working local storage implementation - or do you mean - that it is a problem with the sandboxed iframe?

@oewaoffice if you can confirm that the "image" only request will work, i ll update the PR to use the proposed solution.

regards
helmut

cramforce commented:
@hjanuschka The problem on iOS is that it seems to work, but when you restart the browser the 3p localStorage is gone.

You can try this instead

```
"oewa": {
    "pageview": "https://krone.oewabox.at/j0=,,,r=${canonicalUrl};+,cp=${title}%2Fmoewa%2F+ssl=1+hn=${canonicalHost};;;?lt=${pageViewId}&x=${screenWidth}x${screenHeight}x24&c=CLIENT_ID(oewa)"
  },
```

The `CLIENT_ID` inserted at the end is exactly how you implement the localStorage equivalent in AMP.

oewaoffice commented:
Problem is, that ÖWA policy demands usage of oewa.js from dispatcher.oewabox.at. A reproduction is not allowed.

cramforce commented:
@oewaoffice Again: AMP has its own rules. It is possible to make an exception, but you will need to provide a timeframe when the solution I outlined above can be permitted. Otherwise AMP will not grant an exception. This is particularly true when there is already a solution present (This was slightly different in the Adobe and IVW cases where `amp-analytics` cannot current satisfy the technical requirements).

hjanuschka commented:
ok so this would require another board round? 😢 

@cramforce  the only concern i see with the `c=CLIENT_ID(oewa)` is the format.


as from the pretified version:
```
 var B = localStorage.getItem("i00");
                    if (B) {
                        return "&c=" + B
                    } else {
                        var C = "0000",
                            E = C + Math.ceil((new Date()).getTime() / 1000).toString(16) + (32768 | Math.random() * 65535).toString(16) + C;
                        localStorage.setItem("i00", E)
                    }
```

i am not sure if the format is kinda coded-information - so i am a bit scared - that sending CLIENT_ID to `c=` - will mess up the analytics on oewa-side?!


so could a possible scenario be:
  * take this merge request
  * @oewaoffice  commits to do the board round - to get rid of the requirement of oewa.js
  * @oewaoffice  will check and verify with their vendor if sending CLIENT_ID to `c=` will blow everthing up - and commit on a change
  * once this is done, i ll happly do another PR to change the implementation to your proposed one!

cramforce commented:
@oewaoffice To help you put this into your rule surface: What @hjanuschka is doing with the updated PR is not reproducing, he is literally creating your new oewa.js file. You have full control.

cramforce commented:
@hjanuschka We should get the commitment and submit both iframePing base solution and direct implementation. Maybe you could run both on your site and compare metrics.

As it stands now, the iframePing based solution will grossly over count unique visitors on iOS (basically reporting daily actives as overall unique users and 30 day actives == daily actives).

hjanuschka commented:
ok - i am preparinng another pr - based on your proposed solution - will submit later this day  

--  
Helmut Januschka
Sent with Airmail  

Am 1. Juni 2016 um 16:41:15, Malte Ubl (notifications@github.com(mailto:notifications@github.com)) schrieb:

>  
>  
> @hjanuschka(https://github.com/hjanuschka) We should get the commitment and submit both iframePing base solution and direct implementation. Maybe you could run both on your site and compare metrics.
>  
>  
> As it stands now, the iframePing based solution will grossly over count unique visitors on iOS (basically reporting daily actives as overall unique users and 30 day actives == daily actives).
>  
>  
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub(https://github.com/ampproject/amphtml/pull/3076#issuecomment-223013963), or mute the thread(https://github.com/notifications/unsubscribe/ACwftmKxlo40R1wFviWrwJ4eof_emR7Hks5qHZoLgaJpZM4IV_V1).
>  
>  
>  
>  
>

hjanuschka commented:
@cramforce - do you preffer the second pr to use oewadirect as an analytics property name? so we van avoid mwrge conflicts if you merge the pingless version? 
> 
> 
> 
> 



or should the new pr stock arround - and once oewa is ready i ll rebase it and remove the iframe version? 

regards

cramforce commented:
@hjanuschka One thing we need to plan for is that we can eventually switch out the iframe ping based implementation without requiring re-tagging from publishers (would be required if there were additional vars we need for direct implementation). Lets submit both together (direct suffix in name sounds good). Please bundle everything in one PR. Feel free to just add stuff to this PR.

hjanuschka commented:
tbh if i bundle oewa and oewadirect in one pr - this pr maybe not able to merge - lets say oewa takes X months to aproove the direct version - publishers may use the direct version already and break oewa rules!

hjanuschka commented:
@cramforce or should we just trust the publishers :see_no_evil: .
anyway give me this night - i ll try to do it without adittional vars so a swap out would be no nightmare

cramforce commented:
We should merge it, so people can start collecting test metrics. Trust is needed anyway, because we reverse engineered it together in 15 minutes and so can everybody else :)

hjanuschka commented:
@cramforce  ok i added the oewadirect (gimme a few more minutes to fix travis - after its finished; i am not sure if i got the tests on the first commit ). - uses the exact same variables as the iframe-ping solution - tested it  locally - once `oewadirect` should be used, we have to rename `oewadirect` to `oewa`, remove the old `oewa` and remove the ping exception.


@oewaoffice 


i tested it on the exact same article (once with oewa, and once with oewa direct):



   * OEWA DIRECT:

```
https://ssl-krone.oewabox.at/j0=,,,r=http%3A%2F%2Fwww.krone.at%2FOesterreich%2FOesterreich_beim_Rauchen_trauriger_Europameister-Wie_reagieren-Story-512594;+,cp=RedCont%2FNachrichten%2FLokaleNachrichten+ssl=1+hn=www.krone.at;;;?lt=4200&x=2560x1440x24&c=amp-xgex--2wtCy5BAIu6YnsbYGDSrKSlJ9ynBy2-a94z-6aHLpzC-owghkp2Da1HfoV
```



  * OEWA Iframe Ping:
```
https://ssl-krone.oewabox.at/j0=,,,;+,amp=1+cp=RedCont%2FNachrichten%2FLokaleNachrichten+host=www.krone.at+path=%2FOesterreich%2FOesterreich_beim_Rauchen_trauriger_Europameister-Wie_reagieren-Story-512594+ssl=1+hn=secure.krone.at;;;?lt=iox7szvm&x=2560x1440x24&c=0000574f2683bb7e0000
```


can you verify that this would work, IF the board allows the missing oewa.js exception?
my biggest fear is the `c=` - with the CLIENT_ID from am its pretty large (`amp-xgex--2wtCy5BAIu6YnsbYGDSrKSlJ9ynBy2-a94z-6aHLpzC-owghkp2Da1HfoV`)

cramforce commented on extensions/amp-analytics/0.1/vendors.js:
```diff
@@ -747,3 +781,10 @@ ANALYTICS_CONFIG['adobeanalytics_nativeConfig']
   ['triggers']['pageLoad']['iframe' +
 /* TEMPORARY EXCEPTION */ 'Ping'] = true;
 
+ANALYTICS_CONFIG['adobeanalytics_nativeConfig']
```
You duped these lines

hjanuschka replied:
ahh thx, stupid `git mergetool` messed it up - fixed

cramforce commented:
Looks great. One comment from me.
@avimehta Could you take a quick look?

Regarding the long client id: We have an open feature request to create a filter syntax to change the format.

hjanuschka commented:
@cramforce comment - fixed.

avimehta commented on extensions/amp-analytics/0.1/vendors.js:
```diff
@@ -507,6 +507,40 @@ export const ANALYTICS_CONFIG = {
       },
     },
   },
+  'oewadirect': {
```
Add a pace before this line. Same for `oewa`

avimehta replied:
I know that other vendors are not sorted, but can you move these so that these are in the right place (alphabetical sorting?) Same for the second config.

avimehta commented on extensions/amp-analytics/0.1/vendors.js:
```diff
@@ -507,6 +507,40 @@ export const ANALYTICS_CONFIG = {
   },
+  'oewadirect': {
+    'vars': {
+      'amp': '1',
```
is this value going to be anything other than 1? If not, maybe put it in the url? same below?

avimehta commented on examples/analytics.amp.html:
```diff
@@ -375,6 +375,22 @@
+</script>
+</amp-analytics>
+<!-- End OEWA example -->
+
```
add an example for oewadirect as well?

avimehta commented on extensions/amp-analytics/amp-analytics.md:
```diff
@@ -164,6 +164,26 @@ Type attribute value: `mediametrie`
 
+### OEWA
+
+Type attribute value: `oewa`
```
Maybe move the last paragraph here? If I were a publisher, I probably would not get to that para to see that there is a faster, (arguably) better alternative available.

avimehta commented:
@hjanuschka Added a few comments. Some of the comments are optional but I would appreciate if you addressed all of them. Overall, this PR looks good now. Will merge once things are addressed.

hjanuschka commented:
@avimehta thx for the review - all comments adressed! 👍

hjanuschka commented:
thx for the merge!!! @cramforce @avimehta  when can we expect the changes to be live availaible?

avimehta commented:
Release usually takes about 1 week or so. If a new build is canaried,
tomorrow, you can expect a new stable version in a week from tomorrow.
(sometimes there can be unforeseen delays though)

On Wed, Jun 1, 2016 at 11:27 PM Helmut Januschka <notifications@github.com>
wrote:

> thx for the merge!!! @cramforce <https://github.com/cramforce> @avimehta
> <https://github.com/avimehta> when can we expect the changes to be live
> availaible?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ampproject/amphtml/pull/3076#issuecomment-223206427>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAK8n39zVg98jt8U4-Sodv4GePeiRXpoks5qHne7gaJpZM4IV_V1>
> .
>

erwinmombay commented:
/cc @sriramkrish85 as he is on release duty this week if this needs to be in tomorrows canary

hjanuschka commented:
@erwinmombay would be nice to get it online as fast as possible ;) 
thank you guys you are doing an awesome job!
